### PR TITLE
FIX: Ensure backwards compatibility with old ENV variables

### PR DESCRIPTION
**Pull request by eviltrout**
Fast Rejection "aka Backscatter" will not execute unless the new
ENV variable is present. The original policy will work with both.

### CODE REVIEW

File: boot [hunk 1]
```diff
@@ -29,13 +29,18 @@ done
...
+for v in DISCOURSE_API_KEY DISCOURSE_API_USERNAME; do
 	if [ -z "${!v}" ]; then
 		echo "FATAL ERROR: $v env var is not set." >&2
 		exit 1
 	fi
 done
 
+if [ -z "$DISCOURSE_BASE_URL" ] && [ -z "$DISCOURSE_MAIL_ENDPOINT" ] ; then
+  echo "FATAL ERROR: You need to define DISCOURSE_BASE_URL or DISCOURSE_MAIL_ENDPOINT" >&2
+  exit 1
```
Comment (mpalmer) [inline boot]: These look an awful lot like spaces, not tabs...

File: discourse-smtp-fast-rejection.rb [hunk 2]
```diff
@@ -13,7 +13,7 @@ def logger
 end
 
 def fatal(*args)
-  logger.crit *args
+  logger.crit(*args)
```
Comment (mpalmer) [inline discourse-smtp-fast-rejection.rb]: Seems unnecessary to change this, but style consistency, I guess... :woman_shrugging:

File: discourse-smtp-fast-rejection.rb [hunk 3]
```diff
@@ -25,7 +25,10 @@ def main
   real_env = JSON.parse(File.read(ENV_FILE))
 
   %w{DISCOURSE_BASE_URL DISCOURSE_API_KEY DISCOURSE_API_USERNAME}.each do |kw|
-    fatal "env var %s is required", kw unless real_env[kw]
+    unless real_env[kw]
+      logger.debug "env variable #{kw} missing, fast rejection disabled"
+      real_env['FAST_REJECTION_DISABLED'] = true
```
Comment (mpalmer) [inline discourse-smtp-fast-rejection.rb]: This smells to me like it's time to put this into a proper class and flag this "lolidunno" mode with an instance variable.  Mangling the parsed environment to signal this feels like kicking a puppy.  A *really* cute puppy.

File: receive-mail [hunk 4]
```diff
@@ -25,10 +25,16 @@ def main
...
 
-	%w{DISCOURSE_BASE_URL DISCOURSE_API_KEY DISCOURSE_API_USERNAME}.each do |kw|
+	%w{DISCOURSE_API_KEY DISCOURSE_API_USERNAME}.each do |kw|
 		fatal "env var %s is required", kw unless real_env[kw]
 	end
 
+  if real_env['DISCOURSE_MAIL_ENDPOINT'].nil? &&
+      real_env['DISCOURSE_BASE_URL'].nil?
+
+		fatal "DISCOURSE_MAIL_ENDPOINT or DISCOURSE_BASE_URL env var missing"
```
Comment (mpalmer) [inline receive-mail]: `s/ or / and /`

File: discourse-smtp-fast-rejection.rb (hunk 2 above)
Comment (eviltrout) [inline discourse-smtp-fast-rejection.rb]: I run a ruby linter locally on save of files and fix issues it pops up. In this case, without the parentheses it's ambiguous about whether you want to multiply or expand args.

**Comment by eviltrout** [comment]
Okay, following the advice to split it into an actual class, I did a big refactor:

- Created a base class for both receivers with common helpers
- The disabled flag is now an instance variable on the FastRejection class
- Added specs for everything.

**Comment by mpalmer** [comment]
OMG specs... :heart_eyes: 

I'll do another round of code review on monday.

**Comment by eviltrout** [comment]
Just a polite reminder I'm still waiting for this review :)

File: discourse-smtp-fast-rejection.rb (hunk 2 above)
Comment (mpalmer) [inline discourse-smtp-fast-rejection.rb]: Ooooooh, yeah... didn't think of that.  Definitely want some parens there, then...

File: discourse-smtp-fast-rejection.rb [hunk 5]
```diff
@@ -1,111 +1,10 @@
 #!/usr/bin/env ruby
```
Comment (mpalmer) [inline discourse-smtp-fast-rejection.rb]: Oh, I forgot to mention this previously... any chance you could rip the `.rb` off the end of this?  Implementation details shouldn't be in program names.

**Comment by mpalmer** [comment]
Congratulations!  You have survived the gauntlet of a womble code review.  :grinning: It all looks good to me.

File: spec/spec_helper.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+  end
+
+	config.include FileHelpers
+
+  config.mock_with :rspec do |mocks|
+    mocks.verify_partial_doubles = true
+  end
+
+  config.shared_context_metadata_behavior = :apply_to_host_groups
+end
```
Comment (eviltrout) [inline spec/spec_helper.rb]: Ah, that's because rspec autogenerated that file with spaces so my editor settings didn't catch it. Thanks.

**Comment by eviltrout** [comment]
@mpalmer Okay I've renamed the file and fixed the spaces in the file @tgxworld pointed out.

Merge if it looks good and I'll tackle the next step on Monday!

File: spec/spec_helper.rb (hunk 6 above)
Comment (tgxworld) [inline spec/spec_helper.rb]: No problem 👍 Spaces and tabs are super easy to spot on Github. 8 spaces for a tab 😟

**Pull request merged**